pkg/kubelet/cm/cpumanager/cpu_assignment.go

```diff
@@ -184,6 +207,18 @@ func (s *socketsFirst) sortAvailableSockets() []int {
 	return sockets
 }
 
+// If sockets  higher in the memory hierarchy than NUMA nodes, then UncoreCache
+// sit directly below NUMA Nodes in the memory hierchy
+func (s *socketsFirst) sortAvailableUnCoreCaches() []int {
+	var result []int
+	for _, uncore := range s.acc.sortAvailableNUMANodes() {
+		uncore := s.acc.details.UnCoreInNUMANodes(uncore).UnsortedList()
+		s.acc.sort(uncore, s.acc.details.CPUsInUnCoreCaches)
+		result = append(result, uncore...)
+	}
+	return result
+}
```

Comment:
uhm, makes sense. Will look again more carefully.

Reply:
So, in both cases `sortAvailableUncoreCaches` uses the same logic but depends on the behavior of `sortAvailableNUMANodes`, and only the latter will change its behavior depending if `numaFirst` or `socketFirst`, correct?
If so, this means that `sortAvailableUncoreCaches` per se should not be in the `numaOrSocketsFirstFuncs` interface, because the logic doesn't change; only the logic of its dependencies do.
This also seems what an admittedly quick re-read of the KEP is confirming (will look carefully again later).

If all the above holds, we can simplify the code like this:
```
diff --git a/pkg/kubelet/cm/cpumanager/cpu_assignment.go b/pkg/kubelet/cm/cpumanager/cpu_assignment.go
index f99d92f7aac..597bb9fd5f3 100644
--- a/pkg/kubelet/cm/cpumanager/cpu_assignment.go
+++ b/pkg/kubelet/cm/cpumanager/cpu_assignment.go
@@ -97,7 +97,6 @@ type numaOrSocketsFirstFuncs interface {
 	takeFullSecondLevel()
 	sortAvailableNUMANodes() []int
 	sortAvailableSockets() []int
-	sortAvailableUncoreCaches() []int
 	sortAvailableCores() []int
 }
 
@@ -119,18 +118,6 @@ func (n *numaFirst) takeFullSecondLevel() {
 	n.acc.takeFullSockets()
 }
 
-// If NUMA nodes are higher in the memory hierarchy than sockets, then just
-// sort the UncoreCache within the socket and return them.
-func (n *numaFirst) sortAvailableUncoreCaches() []int {
-	var result []int
-	for _, socket := range n.acc.sortAvailableNUMANodes() {
-		uncore := n.acc.details.UncoreInNUMANodes(socket).UnsortedList()
-		n.acc.sort(uncore, n.acc.details.CPUsInUncoreCaches)
-		result = append(result, uncore...)
-	}
-	return result
-}
-
 // If NUMA nodes are higher in the memory hierarchy than sockets, then just
 // sort the NUMA nodes directly, and return them.
 func (n *numaFirst) sortAvailableNUMANodes() []int {
@@ -197,18 +184,6 @@ func (s *socketsFirst) sortAvailableSockets() []int {
 	return sockets
 }
 
-// If sockets higher in the memory hierarchy than NUMA nodes, then UncoreCache
-// sit directly below NUMA Nodes in the memory hierchy
-func (s *socketsFirst) sortAvailableUncoreCaches() []int {
-	var result []int
-	for _, uncore := range s.acc.sortAvailableNUMANodes() {
-		uncore := s.acc.details.UncoreInNUMANodes(uncore).UnsortedList()
-		s.acc.sort(uncore, s.acc.details.CPUsInUncoreCaches)
-		result = append(result, uncore...)
-	}
-	return result
-}
-
 // If sockets are higher in the memory hierarchy than NUMA nodes, then cores
 // sit directly below NUMA Nodes in the memory hierarchy.
 func (s *socketsFirst) sortAvailableCores() []int {
@@ -377,10 +352,20 @@ func (a *cpuAccumulator) freeSockets() []int {
 	return free
 }
 
+func (a *cpuAccumulator) sortAvailableUncoreCaches() []int {
+	var result []int
+	for _, numa := range a.sortAvailableNUMANodes() {
+		uncore := a.details.UncoreInNUMANodes(numa).UnsortedList()
+		a.sort(uncore, a.details.CPUsInUncoreCaches)
+		result = append(result, uncore...)
+	}
+	return result
+}
+
 // Returns free UncoreCache IDs as a slice sorted by sortAvailableUnCoreCache().
 func (a *cpuAccumulator) freeUncoreCache() []int {
 	free := []int{}
-	for _, uncore := range a.numaOrSocketsFirst.sortAvailableUncoreCaches() {
+	for _, uncore := range a.sortAvailableUncoreCaches() {
 		if a.isUncoreCacheFree(uncore) {
 			free = append(free, uncore)
 		}
@@ -573,7 +558,7 @@ func (a *cpuAccumulator) takeFullUnCore() {
 // First try to take full UncoreCache, if available and need is at least the size of the UncoreCache group.
 // Second try to take the partial UncoreCache if available and the request size can fit w/in the UncoreCache.
 func (a *cpuAccumulator) takeUncoreCache() {
-	for _, uncore := range a.numaOrSocketsFirst.sortAvailableUncoreCaches() {
+	for _, uncore := range a.sortAvailableUncoreCaches() {
 		// take full UncoreCache if the CPUs needed is greater than free UncoreCache size
 		if a.numCPUsNeeded >= a.topo.NumCPUs/a.topo.NumUncoreCache {
 			a.takeFullUnCore()
```

note thispatch  assumes both my other simplifications proposals have been applied, so we don't have `takeThirdLevel` anymore.

Reply:
This makes sense, applied these changes to have one `sortAvailableUncoreCaches` by NUMA.
Removed takeThirdLevel as well to simplify.